### ISSUE TITLE
hook-pyfiglet.py

### ISSUE THREAD

**Issue by DeflatedPickle** [pyinstaller/pyinstaller#2389]
```python
from PyInstaller.utils.hooks import collect_data_files

hiddenimports = ["pyfiglet.fonts"]

datas = collect_data_files ("pyfiglet")
````

**Comment by htgoebel** [comment]
Thanks for this hook. Would you please send a pull-request? And please mind adding
*  the copyright header you can see in the other hooks (mind that we have 2017 now :-) 
* a comment with the url to `pyfiglet` and 
* which version you have tested it with.

Thanks.

**Comment by nejch** [comment]
@DeflatedPickle did you actually get this hook to work? I don't think it does with current versions of pyinstaller and pyfiglet. The package uses `pkg_resources` to load fonts, resulting in errors similar to #1898:
```
    f = Figlet(font='slant')
  File "lib\site-packages\pyfiglet\__init__.py", line 710, in __init__
  File "lib\site-packages\pyfiglet\__init__.py", line 717, in setFont
  File "lib\site-packages\pyfiglet\__init__.py", line 90, in __init__
  File "lib\site-packages\pyfiglet\__init__.py", line 100, in preloadFont
  File "lib\site-packages\pkg_resources\__init__.py", line 1191, in resource_exists
  File "lib\site-packages\pkg_resources\__init__.py", line 1459, in has_resource
  File "lib\site-packages\pkg_resources\__init__.py", line 1509, in _has
NotImplementedError: Can't perform this operation for unregistered loader type
```

 So I think this depends on #2748 unless you provide the font parameter as a filename to avoid using pkg_resources - see below the relevant code for reference: 
```python
def preloadFont(cls, font):
        """
        Load font data if exist
        """
        for extension in ('tlf', 'flf'):
            fn = '%s.%s' % (font, extension)
            if pkg_resources.resource_exists('pyfiglet.fonts', fn):
                data = pkg_resources.resource_string('pyfiglet.fonts', fn)
                data = data.decode('UTF-8', 'replace')
                return data
            elif os.path.isfile(font):
                with open(font, 'rb') as f:
                    return f.read().decode('UTF-8', 'replace')
        else:
            raise FontNotFound(font)
```

**Comment by ossie-git** [comment]
Was anyone able to get scripts using pyfiglet to work?  I tried the above hook which didn't work and then tried updating my spec to include the font files as shown below:

```
 datas=[('/usr/local/lib/python2.7/site-packages/pyfiglet/fonts','.')],
```

It creates the executable but when I run it, I get the following:

```
Traceback (most recent call last):
  File "app.py", line 242, in <module>
    cprint(figlet_format(banner, font="standard"),"blue")
  File "site-packages/pyfiglet/__init__.py", line 43, in figlet_format
  File "site-packages/pyfiglet/__init__.py", line 710, in __init__
  File "site-packages/pyfiglet/__init__.py", line 717, in setFont
  File "site-packages/pyfiglet/__init__.py", line 90, in __init__
  File "site-packages/pyfiglet/__init__.py", line 100, in preloadFont
  File "site-packages/pkg_resources/__init__.py", line 1161, in resource_exists
  File "site-packages/pkg_resources/__init__.py", line 419, in get_provider
ImportError: No module named fonts
Failed to execute script app
```

I'm kind of surprised that this isn't working considering that a lot of scripts make use of pyfiglet.  Thanks.

**Comment by bjones1** [comment]
Does this works?

``` Python
from PyInstaller.utils.hooks import collect_all

datas, binaries, hiddenimports = collect_all("pyfiglet")
```

**Comment by SeeDoubleYou** [comment]
> Does this work?
> 
> ```python
> from PyInstaller.utils.hooks import collect_all
> 
> datas, binaries, hiddenimports = collect_all("pyfiglet")
> ```

No, unfortunately it doesn't. This still results in a ModuleNotFoundError on pyfiglet.fonts

**Comment by bjones1** [comment]
Hmm. Try running PyInstaller with [--log-level=DEBUG](https://pyinstaller.readthedocs.io/en/stable/usage.html#general-options) and [--debug=all](https://pyinstaller.readthedocs.io/en/stable/usage.html#how-to-generate). Look at the output of the build to see if pyfiglet files are picked up. Look at the files in `dist/` to see what's missing.

**Comment by htgoebel** [comment]
>   File "site-packages/pyfiglet/__init__.py", line 100, in preloadFont

Look at this line, what is done there.

**Comment by caramelomartins** [comment]
I've found a workround that has worked for me and might work for other people too, so I'd like to share it. I've added the following as a PyInstaller hook (`hook-pyfiglet.py`): 

> Does this work?
> 
> ```python
> from PyInstaller.utils.hooks import collect_all
> 
> datas, binaries, hiddenimports = collect_all("pyfiglet")
> ```

This seems to have resolved the import issues, as expected, but raised the error `NotImplementedError: Can't perform this operation for unregistered loader type`, as said in https://github.com/pyinstaller/pyinstaller/issues/2389#issuecomment-420941860.

I've followed what has been discussed in #2748 and understood that the issue was now related to the fact that a runtime hook was registering the `pkg_resources.NullProvider`. Since adding another runtime hook wouldn't resolve the problem, as the installed ones would have processed afterwards, I ended up adding the following snippet to my script:

```
import pkg_resources
import sys

from pyimod03_importers import FrozenImporter

if getattr(sys, 'frozen', False):
   pkg_resources.register_loader_type(FrozenImporter, pkg_resources.DefaultProvider)
```

In essence, instead of registering `NullProvider`, we now register `FrozenImporter` with `DefaultProvider`, which inherits from `NullProvider` anyway.

Now, this works for me but I cannot guarantee this doesn't have side effects, **use with caution please**. Nonetheless, since it has worked for me and the issues haven't been fixed yet, I thought someone might have been interested in a workaround.

I have noticed this issue is dependent on #2748 , which is itself dependent on another issue. I'm not sure if this workaround helps in figuring out a way of solving any of those issues but if there's any approach you think could be taken, I don't mind trying to help.

**Comment by miniluigi008** [comment]
I understand there is a different issue at hand here, but for anyone looking to quickly use pyfiglet with pyinstaller, doesn't this "just work" by adding a few things to the `app.spec`? (It seems to work for me.)
```
hiddenimports=['pyfiglet.fonts'],
datas=[('[insert full path to site packages]/pyfiglet/fonts/standard.flf','pyfiglet/fonts')],
```
Good luck on the issue everyone.

**Comment by Thonleb** [comment]
> J'ai trouvé une solution de contournement qui a fonctionné pour moi et pourrait également fonctionner pour d'autres personnes, alors j'aimerais la partager. J'ai ajouté ce qui suit en tant que hook PyInstaller ( `hook-pyfiglet.py`):
> 
> > Est-ce que ça marche?
> > ```python
> > de  PyInstaller . utils . crochets  import  collect_all
> > 
> > datas , binaries , hiddenimports  =  collect_all ( "pyfiglet" )
> > ```
> 
> Cela semble avoir résolu les problèmes d'importation, comme prévu, mais a soulevé l'erreur `NotImplementedError: Can't perform this operation for unregistered loader type`, comme indiqué dans [# 2389 (commentaire)](https://github.com/pyinstaller/pyinstaller/issues/2389#issuecomment-420941860) .
> 
> J'ai suivi ce qui a été discuté dans [# 2748](https://github.com/pyinstaller/pyinstaller/issues/2748) et j'ai compris que le problème était maintenant lié au fait qu'un hook d'exécution enregistrait le `pkg_resources.NullProvider`. Étant donné que l'ajout d'un autre hook d'exécution ne résoudrait pas le problème, car ceux installés auraient été traités par la suite, j'ai fini par ajouter l'extrait de code suivant à mon script:
> 
> ```
> import pkg_resources
> import sys
> 
> from pyimod03_importers import FrozenImporter
> 
> if getattr(sys, 'frozen', False):
>    pkg_resources.register_loader_type(FrozenImporter, pkg_resources.DefaultProvider)
> ```
> 
> Essentiellement, au lieu de nous inscrire `NullProvider`, nous nous inscrivons maintenant `FrozenImporter`avec `DefaultProvider`, qui hérite de `NullProvider`toute façon.
> 
> Maintenant, cela fonctionne pour moi mais je ne peux pas garantir que cela n'a pas d'effets secondaires, **utilisez avec prudence s'il vous plaît** . Néanmoins, comme cela a fonctionné pour moi et que les problèmes n'ont pas encore été résolus, j'ai pensé que quelqu'un aurait pu être intéressé par une solution de contournement.
> 
> J'ai remarqué que ce problème dépend de [# 2748](https://github.com/pyinstaller/pyinstaller/issues/2748) , qui lui-même dépend d'un autre problème. Je ne sais pas si cette solution de contournement aide à trouver un moyen de résoudre l'un de ces problèmes, mais s'il y a une approche qui, selon vous, pourrait être adoptée, cela ne me dérange pas d'essayer d'aider.



> I've found a workround that has worked for me and might work for other people too, so I'd like to share it. I've added the following as a PyInstaller hook (`hook-pyfiglet.py`):
> 
> > Does this work?
> > ```python
> > from PyInstaller.utils.hooks import collect_all
> > 
> > datas, binaries, hiddenimports = collect_all("pyfiglet")
> > ```
> 
> This seems to have resolved the import issues, as expected, but raised the error `NotImplementedError: Can't perform this operation for unregistered loader type`, as said in [#2389 (comment)](https://github.com/pyinstaller/pyinstaller/issues/2389#issuecomment-420941860).
> 
> I've followed what has been discussed in #2748 and understood that the issue was now related to the fact that a runtime hook was registering the `pkg_resources.NullProvider`. Since adding another runtime hook wouldn't resolve the problem, as the installed ones would have processed afterwards, I ended up adding the following snippet to my script:
> 
> ```
> import pkg_resources
> import sys
> 
> from pyimod03_importers import FrozenImporter
> 
> if getattr(sys, 'frozen', False):
>    pkg_resources.register_loader_type(FrozenImporter, pkg_resources.DefaultProvider)
> ```
> 
> In essence, instead of registering `NullProvider`, we now register `FrozenImporter` with `DefaultProvider`, which inherits from `NullProvider` anyway.
> 
> Now, this works for me but I cannot guarantee this doesn't have side effects, **use with caution please**. Nonetheless, since it has worked for me and the issues haven't been fixed yet, I thought someone might have been interested in a workaround.
> 
> I have noticed this issue is dependent on #2748 , which is itself dependent on another issue. I'm not sure if this workaround helps in figuring out a way of solving any of those issues but if there's any approach you think could be taken, I don't mind trying to help.



> I've found a workround that has worked for me and might work for other people too, so I'd like to share it. I've added the following as a PyInstaller hook (`hook-pyfiglet.py`):
> 
> > Does this work?
> > ```python
> > from PyInstaller.utils.hooks import collect_all
> > 
> > datas, binaries, hiddenimports = collect_all("pyfiglet")
> > ```
> 
> This seems to have resolved the import issues, as expected, but raised the error `NotImplementedError: Can't perform this operation for unregistered loader type`, as said in [#2389 (comment)](https://github.com/pyinstaller/pyinstaller/issues/2389#issuecomment-420941860).
> 
> I've followed what has been discussed in #2748 and understood that the issue was now related to the fact that a runtime hook was registering the `pkg_resources.NullProvider`. Since adding another runtime hook wouldn't resolve the problem, as the installed ones would have processed afterwards, I ended up adding the following snippet to my script:
> 
> ```
> import pkg_resources
> import sys
> 
> from pyimod03_importers import FrozenImporter
> 
> if getattr(sys, 'frozen', False):
>    pkg_resources.register_loader_type(FrozenImporter, pkg_resources.DefaultProvider)
> ```
> 
> In essence, instead of registering `NullProvider`, we now register `FrozenImporter` with `DefaultProvider`, which inherits from `NullProvider` anyway.
> 
> Now, this works for me but I cannot guarantee this doesn't have side effects, **use with caution please**. Nonetheless, since it has worked for me and the issues haven't been fixed yet, I thought someone might have been interested in a workaround.
> 
> I have noticed this issue is dependent on #2748 , which is itself dependent on another issue. I'm not sure if this workaround helps in figuring out a way of solving any of those issues but if there's any approach you think could be taken, I don't mind trying to help.

**"Now, this works for me but I cannot guarantee this doesn't have side effects, use with caution please. "**

Exactly, it fixed my problem thx, but when I use sys.exit(), it doesn't really work but it totally bugs my code, do you have a fix?

**Comment by caramelomartins** [comment]
Sorry @Thonleb, I don't have a solution for you. Since this is dependent on a few outstanding issues to be properly solved, I guess you can only get something working (hacked) by trial and error. Do you have more details about what "it totally bugs my code" means?

**Comment by vaibhavard** [comment]
@benduckwdesign , great suggestion... My issue was resolved by this!
